libsql-server/src/replication/snapshot_store.rs

```diff
@@ -0,0 +1,333 @@
+            end_frame_no: 0,
+            last_seen_frame_no: None,
+            db_size,
+            name: namespace_name,
+        })
+    }
+
+    /// Returns the biggest snapshot for namespace that contains frame_no
+    // FIXME: make async?
+    pub fn find(&self, namespace: &NamespaceName, frame_no: FrameNo) -> Result<Option<Uuid>> {
+        let conn = self.inner.conn.lock();
+        let mut stmt = conn.prepare_cached(
+            r#"SELECT snapshot_id FROM snapshots 
+            WHERE namespace = :namespace
+                AND start_frame_no <= :fno
+                AND end_frame_no >= :fno
+            ORDER BY (end_frame_no - start_frame_no) DESC
+            LIMIT 1"#,
+        )?;
+        let snapshot_id = stmt
+            .query_row(
+                named_params!( ":namespace": namespace.as_str(), ":fno": frame_no),
+                |row| {
+                    let s = row.get_ref(0)?.as_str()?;
+                    Ok(Uuid::from_str(s).unwrap())
+                },
+            )
+            .optional()?;
+
+        Ok(snapshot_id)
+    }
+
+    pub async fn find_file(
+        &self,
+        namespace: &NamespaceName,
+        frame_no: FrameNo,
+    ) -> Result<Option<SnapshotFile>> {
+        match self.find(namespace, frame_no)? {
+            Some(snapshot_id) => {
+                let path = self
+                    .inner
+                    .snapshots_path(namespace)
+                    .join(snapshot_id.to_string());
+                let file = SnapshotFile::open(path).await?;
+                Ok(Some(file))
+            }
+            None => Ok(None),
+        }
+    }
+
+    pub async fn delete_all(&self, namespace: NamespaceName) -> Result<()> {
+        let inner = self.inner.clone();
+        tokio::task::spawn_blocking(move || inner.delete_all(&namespace))
+            .await
+            .unwrap()?;
+
+        Ok(())
+    }
+}
+
+pub struct SnapshotBuilder {
+    /// Temporary file to hold the snapshot, before it's persisted with as self.name
+    snapshot_file: NamedTempFile,
+    store: Arc<SnapshotStoreInner>,
+    frame_count: u64,
+    end_frame_no: FrameNo,
+    last_seen_frame_no: Option<FrameNo>,
+    db_size: u32,
+    name: NamespaceName,
+}
+
+impl SnapshotBuilder {
+    pub fn add_frame(&mut self, frame: &FrameBorrowed) -> Result<()> {
+        let frame_no = frame.header().frame_no.get();
+        match self.last_seen_frame_no {
+            Some(last_seen) => {
+                assert!(last_seen > frame_no);
+            }
+            None => {
+                self.end_frame_no = frame_no;
+            }
+        }
+
+        self.last_seen_frame_no = Some(frame_no);
+        self.frame_count += 1;
+
+        self.snapshot_file.write_all(frame.as_bytes()).unwrap();
+
+        Ok(())
+    }
+
+    pub fn finish(mut self, start_frame_no: FrameNo) -> Result<()> {
+        self.snapshot_file
+            .seek(std::io::SeekFrom::Start(0))
+            .unwrap();
```

Comment:
should we handle this suprrious io errors dont sound like fun

Reply:
we sure should, but for now I'm only interested in having the skeleton code in tree, so that I can iron out the storage layer with @Horusiath